modules/DOMStateStorage.js

```diff
@@ -8,18 +8,33 @@ function createKey(key) {
   return KeyPrefix + key
 }
 
+function isHistoryKey(key) {
+  return key.indexOf(KeyPrefix) === 0
+}
+
 export function saveState(key, state) {
   try {
     window.sessionStorage.setItem(createKey(key), JSON.stringify(state))
   } catch (error) {
-    if (error.name === QuotaExceededError || window.sessionStorage.length === 0) {
-      // Probably in Safari "private mode" where sessionStorage quota is 0. #42
-      warning(
-        false,
-        '[history] Unable to save state; sessionStorage is not available in Safari private mode'
-      )
-
-      return
+    if (error.name === QuotaExceededError) {
+      if (window.sessionStorage.length === 0) {
+        // Probably in Safari "private mode" where sessionStorage quota is 0. #42
+        warning(
+          false,
+          '[history] Unable to save state; sessionStorage is not available in Safari private mode'
+        )
+
+        return
+      }
+
+      // Reached quota, remove first (oldest) item in storage
+      for (const storedKey in window.sessionStorage) {
```

Comment:
> Since we can't rely on browsers for consistent behavior here, one idea I had was that we could keep our own array in memory, ordered however we want (probably by least recently used) and then just try to persist that array to sessionStorage each time...

Originally I was going to go with this idea. Before I do, there's also this method here on the storage object: https://developer.mozilla.org/en-US/docs/Web/API/Storage/key which from my testing seems to always return the correct key - ie `.key(0)` will return the oldest key in storage.